### PR TITLE
Develop 2.9.3

### PR DESCRIPTION
**Pull request by zebra-lucky**
Merge electrum-dash with electrum 2.9.3

### CODE REVIEW

File: RELEASE-NOTES [hunk 1]
```diff
@@ -1,3 +1,226 @@
...
+    commands can use them. Wallets can be loaded using: 'electrum
+    daemon load_wallet [-w path]'. This command will require a
+    password if the wallet is encrypted.
+  * Invoices and contacts are stored in the wallet file and are no
+    longer shared between wallets. Previously created invoices and
+    contacts files may be imported from the menu.
+  * Fees improvements:
+    - Dynamic fees are enabled by default.
+    - Child Pays For Parent (CPFP) dialog in the GUI.
+    - RBF is automatically proposed for low fee transactions.
```
Comment (akhavr) [inline RELEASE-NOTES]: There's no RBF in dash.  Pls check and remove the respective code.

File: RELEASE-NOTES [hunk 2]
```diff
@@ -1,3 +1,226 @@
...
+# Release 2.7.16
+  * add Testnet support (fix #541)
+  * allow daemon to be launched in the foreground (fix #1873)
+  * Qt: use separate tabs for addresses and UTXOs
+  * Qt: update fee slider with a network callback
+  * Ledger: new ui and mobile 2fa validation (neocogent)
+
+# Release 2.7.15
+  * Use fee slider for both static and dynamic fees.
+  * Add fee slider to RBF dialog (fix #2083).
```
Comment (akhavr) [inline RELEASE-NOTES]: RBF

File: RELEASE-NOTES [hunk 3]
```diff
@@ -1,3 +1,226 @@
...
+   to store private keys. Therefore, so-called 'hardware wallets' will
+   be referred to as 'hardware keystores'.
+ * Hardware keystores:
+   - The Ledger Nano S is supported.
+   - Hardware keystores can be used as cosigners in multi-signature
+     wallets.
+   - Multiple hardware cosigners can be used in the same multisig
+     wallet. One icon per keystore is displayed in the satus bar. Each
+     connected device will co-sign the transaction.
+ * Replace-By-Fee: RBF transactions are supported in both Qt and
```
Comment (akhavr) [inline RELEASE-NOTES]: RBF

File: lib/synchronizer.py [hunk 4]
```diff
@@ -124,7 +136,7 @@ def tx_response(self, response):
         if not params:
             return
         tx_hash, tx_height = params
-        assert tx_hash == hash_encode(Hash(result.decode('hex')))
+        #assert tx_hash == hash_encode(Hash(result.decode('hex')))
```
Comment (akhavr) [inline lib/synchronizer.py]: commented out in the original code?

File: lib/tests/test_account.py [hunk 5]
```diff
@@ -1,123 +1,9 @@
...
-                expected = v[label][n]
-                self.assertEquals(expected, a.derive_pubkey_from_xpub(a.xpub, for_change, n))
-                self.assertEquals(expected, a.get_pubkey_from_xpub(a.xpub, for_change, n))
-                self.assertEquals(expected, a.derive_pubkeys(for_change, n))
-
-                pubkey, = a.get_xpubkeys(for_change, n)
-                xpub, seq = a.parse_xpubkey(pubkey)
-                self.assertEquals(xpub, a.xpub)
-                self.assertEquals(seq, [for_change, n])
+        self.assertTrue(True)
```
Comment (akhavr) [inline lib/tests/test_account.py]: Шикарні тести!

File: lib/network.py [hunk 6]
```diff
@@ -638,24 +681,35 @@ def connection_down(self, server):
...
+        interface.mode = 'default'
+        interface.request = None
+        self.interfaces[server] = interface
+        self.queue_request('server.version',
+                           [ELECTRUM_VERSION, PROTOCOL_VERSION], interface)
         self.queue_request('blockchain.headers.subscribe', [], interface)
         if server == self.default_server:
             self.switch_to_interface(server)
-        self.notify('interfaces')
+        #self.notify('interfaces')
```
[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/network.py]: Yes, in electrum 2.9.3 line is commented.

File: lib/network.py [hunk 7]
```diff
@@ -691,91 +745,178 @@ def maintain_sockets(self):
...
+            self.request_chunk(interface, index+1)
+        else:
+            interface.request = None
+            interface.mode = 'default'
+            interface.print_error('catch up done', interface.blockchain.height())
+            interface.blockchain.catch_up = None
+        self.notify('updated')
+
+    def request_header(self, interface, height):
+        #interface.print_error("requesting header %d" % height)
```
[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/network.py]: Yes, in electrum 2.9.3 line is commented.

File: lib/tests/test_account.py (hunk 5 above)
Comment (zebra-lucky) [reply lib/tests/test_account.py]: )

Comment (zebra-lucky) [reply lib/tests/test_account.py]: То чомусь не прибрали тести для lib/accounts.py
Можна мабуть видалити буде.

File: lib/synchronizer.py (hunk 4 above)
Comment (zebra-lucky) [reply lib/synchronizer.py]: Yes, in electrum 2.9.3 line is commented.

File: lib/blockchain.py [hunk 8]
```diff
@@ -275,44 +379,32 @@ def get_target(self, height, chain=None):
...
+        except:
             return False
+        return True
 
     def connect_chunk(self, idx, hexdata):
         try:
             data = hexdata.decode('hex')
             self.verify_chunk(idx, data)
-            self.print_error("validated chunk %d" % idx)
+            #self.print_error("validated chunk %d" % idx)
```
[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/blockchain.py]: Yes, in electrum 2.9.3 line is commented.

File: lib/blockchain.py [hunk 9]
```diff
@@ -63,144 +66,247 @@ def target_to_bits(target):
...
     def verify_header(self, header, prev_header, bits, target):
-        prev_hash = self.hash_header(prev_header)
-        assert prev_hash == header.get('prev_block_hash'), "prev hash mismatch: %s vs %s" % (prev_hash, header.get('prev_block_hash'))
+        prev_hash = hash_header(prev_header)
+        _hash = hash_header(header)
+        if prev_hash != header.get('prev_block_hash'):
+            raise BaseException("prev hash mismatch: %s vs %s" % (prev_hash, header.get('prev_block_hash')))
+        if bitcoin.TESTNET:
+            return
+        #if bits != header.get('bits'):
```
[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/blockchain.py]: Similar code was deleted in electrum-dash 2.6.4 diff to electrum 2.6.4
Headers not syncing if uncomment this.

[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/blockchain.py]: This was deleted in electrum-dash 2.6.4 relative to electrum 2.6.4

```
-        assert bits == header.get('bits'), "bits mismatch: %s vs %s" % (bits, header.get('bits'))
-        _hash = self.hash_header(header)
-        assert int('0x' + _hash, 16) <= target, "insufficient proof of work: %s vs target %s" % (int('0x' + _hash, 16), target)
```

File: lib/bitcoin.py [hunk 10]
```diff
@@ -419,9 +480,18 @@ def msg_magic(message):
     return "\x19DarkCoin Signed Message:\n" + encoded_varint + message
```
[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/bitcoin.py]: Changed.

File: lib/tests/test_account.py (hunk 5 above)
Comment (zebra-lucky) [reply lib/tests/test_account.py]: Deleted.

File: lib/bitcoin.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zebra-lucky) [reply lib/bitcoin.py]: ```const string strMessageMagic = "DarkCoin Signed Message:\n";``` Defined in dash `src/validation.cpp`. Returned back DarkCoin.

**Pull request merged**